net_util/src/queue_pair.rs

```diff
@@ -193,6 +241,30 @@ impl RxVirtio {
                 0
             };
 
+            if let Some(rate_limiter) = rate_limiter {
+                let mut rate_limit_reached = false;
+
+                if !rate_limiter.consume(1, TokenType::Ops) {
+                    rate_limit_reached = true;
+                } else if !rate_limiter.consume(len as u64, TokenType::Bytes) {
+                    // Revert the OPS consume().
+                    rate_limiter.manual_replenish(1, TokenType::Ops);
+                    rate_limit_reached = true;
+                }
+
+                if rate_limit_reached {
+                    // Set the deferred frame which will be processed when
+                    // the 'rx_rate_limiter' is unblocked
+                    *rx_deferred_frame = Some(len);
```

Comment:
I don't understand the need for introducing a `rx_deferred_frame` and the whole logic that comes with it.
I was expecting that if we reached the rate limiter limit, then we go to the previous descriptor and we're done with it. It's important to unregister the TAP fd after we did so, but nothing else.
Is there an aspect that I'm missing and that's mandatory to make the rate limiter work? I'm asking because this adds a lot of complexity to the code.

Reply:
We need the `rx_deferred_frame` as we can't revert the read from the TAP when the rate limit is reached. Note that we are only able to check whether a RX frame is over the rate limit after the `readv` returns the actual `bytes` being read from the TAP.